js_modules/dagit/packages/core/src/app/ExecutionSessionStorage.tsx

```diff
@@ -134,74 +122,94 @@ function getStorageDataForNamespace(namespace: string, initial: Partial<IExecuti
   }
 
-  _data = data;
-  _dataNamespace = namespace;
-
   return data;
-}
+};
 
-function writeStorageDataForNamespace(namespace: string, data: IStorageData) {
-  _data = data;
-  _dataNamespace = namespace;
-  window.localStorage.setItem(getKey(namespace), JSON.stringify(data));
-}
+const writeStorageDataForKey = (key: string, data: IStorageData) => {
+  window.localStorage.setItem(key, JSON.stringify(data));
+};
+
+export const makeKey = (basePath: string, repoAddress: RepoAddress, pipelineOrJobName: string) =>
+  `dagit.v2.${basePath}-${repoAddress.location}-${repoAddress.name}-${pipelineOrJobName}`;
+
+export const makeOldKey = (repoAddress: RepoAddress, pipelineOrJobName: string) =>
+  `dagit.v2.${repoAddress.name}.${pipelineOrJobName}`;
 
-/* React hook that provides local storage to the caller. A previous version of this
-loaded data into React state, but changing namespaces caused the data to be out-of-sync
-for one render (until a useEffect could update the data in state). Now we keep the
-current localStorage namespace in memory (in _data above) and React keeps a simple
-version flag it can use to trigger a re-render after changes are saved, so changing
-namespaces changes the returned data immediately.
-*/
+/**
+ * React hook that provides local storage to the caller. A previous version of this
+ * loaded data into React state, but changing namespaces caused the data to be out-of-sync
+ * for one render (until a useEffect could update the data in state). Now we keep the
+ * current localStorage namespace in memory (in _data above) and React keeps a simple
+ * version flag it can use to trigger a re-render after changes are saved, so changing
+ * namespaces changes the returned data immediately.
+ */
 export function useExecutionSessionStorage(
-  repositoryName: string,
-  pipelineName: string,
+  repoAddress: RepoAddress,
+  pipelineOrJobName: string,
   initial: Partial<IExecutionSession> = {},
 ): StorageHook {
-  const namespace = `${repositoryName}.${pipelineName}`;
-  const [version, setVersion] = React.useState<number>(0);
+  const {basePath} = React.useContext(AppContext);
 
-  const onSave = (newData: IStorageData) => {
-    writeStorageDataForNamespace(namespace, newData);
-    setVersion(version + 1); // trigger a React render
-  };
+  const key = makeKey(basePath, repoAddress, pipelineOrJobName);
+  const oldKey = makeOldKey(repoAddress, pipelineOrJobName);
+
+  const validate = React.useCallback((value) => value, []);
+
+  const [_, setNewState] = useStateWithStorage(key, validate);
+  const [oldState, setOldState] = useStateWithStorage(oldKey, validate);
 
-  return [getStorageDataForNamespace(namespace, initial), onSave];
+  React.useEffect(() => {
+    if (oldState) {
+      setNewState(oldState);
+
+      // Delete data at old key.
+      setOldState(undefined);
+    }
+  }, [oldState, setNewState, setOldState]);
+
+  return [getStorageDataForNamespace(key, initial), setNewState];
```

Comment:
Hey! I think we need to avoid calling getStorageDataForNamespace() on every render because it returns a new object each time, but also because it does a `JSON.parse(localStorage.getItem())` on every render and it could contain 50kb of YAML or something :-)

I think if we use the useStateWithStorage hook fully, we should be able to do this with just two changes:

- On line 158, let's set _ to newState and return `newState` here instead of the getter.

- The `validate` method receives the raw JSON and is supposed to return the "hydrated value" (with any sort of validation and other rules applied). I think that in our case, that's almost what `getStorageDataForNamespace` does, except that it calls `getJSONForKey(key)` on it's own. If we change `getStorageDataForNamespace` to `validateDataForNamespace` and have it take the JSON in localstorage as it's only param, I think the rest of it is good to go.

I think we'll have to put `initial` into the deps array on 156, so we may need to check that that value is coming in memoized 😬